### PR TITLE
Add DigitalOcean oAuth Provider

### LINKED ISSUE #652: Add DigitalOcean provider
Please provide a DigitalOcean provider: https://www.digitalocean.com/community/tutorials/how-to-use-oauth-authentication-with-digitalocean-as-a-user-or-developer

Alternative is to implement oAuth myself.

### PR DESCRIPTION
**Pull request by ElanHasson**
Adds the DigitalOcean provider.

Resolves #652

### CODE REVIEW

File: NuGet.config [hunk 1]
```diff
@@ -1,11 +1,12 @@
-<?xml version="1.0" encoding="utf-8"?>
```
Comment (martincostello) [inline NuGet.config]: Please revert the changes to this file.

File: src/AspNet.Security.OAuth.DigitalOcean/AspNet.Security.OAuth.DigitalOcean.csproj [hunk 2]
```diff
@@ -0,0 +1,18 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <TargetFrameworks>$(DefaultNetCoreTargetFramework)</TargetFrameworks>
+  </PropertyGroup>
```
Comment (martincostello) [inline src/AspNet.Security.OAuth.DigitalOcean/AspNet.Security.OAuth.DigitalOcean.csproj]:
```suggestion
  <PropertyGroup>
    <PackageValidationBaselineVersion>6.0.4</PackageValidationBaselineVersion>
    <TargetFrameworks>$(DefaultNetCoreTargetFramework)</TargetFrameworks>
  </PropertyGroup>

  <!-- TODO Remove once published to NuGet.org -->
  <PropertyGroup>
    <DisablePackageBaselineValidation>true</DisablePackageBaselineValidation>
  </PropertyGroup>
```

File: src/AspNet.Security.OAuth.DigitalOcean/AspNet.Security.OAuth.DigitalOcean.csproj [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+
+  <PropertyGroup>
+    <Description>ASP.NET Core security middleware enabling DigitalOcean authentication.</Description>
+    <Authors>Elan Hasson</Authors>
+    <PackageTags>DigitalOcean;aspnetcore;authentication;oauth;security</PackageTags>
```
Comment (martincostello) [inline src/AspNet.Security.OAuth.DigitalOcean/AspNet.Security.OAuth.DigitalOcean.csproj:16]:
```suggestion
    <PackageTags>digitalocean;aspnetcore;authentication;oauth;security</PackageTags>
```

File: src/AspNet.Security.OAuth.DigitalOcean/DigitalOceanAuthenticationHandler.cs [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+        var user = tokens.Response!.RootElement.GetProperty("info");
+
+        using var request = new HttpRequestMessage(HttpMethod.Get, Options.UserInformationEndpoint);
+        request.Headers.Accept.Add(new MediaTypeWithQualityHeaderValue("application/json"));
+        request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", tokens.AccessToken);
+
+        using var response = await Backchannel.SendAsync(request, HttpCompletionOption.ResponseHeadersRead, Context.RequestAborted);
+        if (!response.IsSuccessStatusCode)
+        {
+            // await Log.UserProfileErrorAsync(Logger, response, Context.RequestAborted);
```
Comment (martincostello) [inline src/AspNet.Security.OAuth.DigitalOcean/DigitalOceanAuthenticationHandler.cs]: Log message missing.

File: src/AspNet.Security.OAuth.DigitalOcean/DigitalOceanAuthenticationHandler.cs [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+    }
+
+    protected override async Task<AuthenticationTicket> CreateTicketAsync(
+        [NotNull] ClaimsIdentity identity,
+        [NotNull] AuthenticationProperties properties,
+        [NotNull] OAuthTokenResponse tokens)
+    {
+        // Note: DigitalOcean doesn't provide a user information endpoint
+        // so we rely on the details sent back in the token request.
+        var user = tokens.Response!.RootElement.GetProperty("info");
```
Comment (martincostello) [inline src/AspNet.Security.OAuth.DigitalOcean/DigitalOceanAuthenticationHandler.cs]: This isn't used until line 92 so doesn't need to be used up here.

File: src/AspNet.Security.OAuth.DigitalOcean/DigitalOceanAuthenticationHandler.cs [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+        return OAuthTokenResponse.Success(payload);
+    }
+
+    protected override async Task<AuthenticationTicket> CreateTicketAsync(
+        [NotNull] ClaimsIdentity identity,
+        [NotNull] AuthenticationProperties properties,
+        [NotNull] OAuthTokenResponse tokens)
+    {
+        // Note: DigitalOcean doesn't provide a user information endpoint
+        // so we rely on the details sent back in the token request.
```
Comment (martincostello) [inline src/AspNet.Security.OAuth.DigitalOcean/DigitalOceanAuthenticationHandler.cs]: This comment is confusing - it says there's no user information endpoint, but then the code issues a request to a user information endpoint. If it isn't a user information endpoint, what is it?

File: src/AspNet.Security.OAuth.DigitalOcean/DigitalOceanAuthenticationHandler.cs [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+        {
+            // await Log.UserProfileErrorAsync(Logger, response, Context.RequestAborted);
+            throw new HttpRequestException("An error occurred while retrieving the user profile.");
+        }
+
+        using var payload = JsonDocument.Parse(await response.Content.ReadAsStringAsync(Context.RequestAborted));
+
+        var account = payload.RootElement.GetProperty("account");
+
+        identity.AddClaim(new Claim("email_verified", account.GetProperty("email_verified").ToString().ToLowerInvariant()));
```
Comment (martincostello) [inline src/AspNet.Security.OAuth.DigitalOcean/DigitalOceanAuthenticationHandler.cs]: The claim name should follow the format we use in the other providers and start with `urn` ([examples](https://github.com/aspnet-contrib/AspNet.Security.OAuth.Providers/search?q=urn)).

Comment (martincostello) [inline src/AspNet.Security.OAuth.DigitalOcean/DigitalOceanAuthenticationHandler.cs]: What is the type of the `email_verified` property? If it's a string, you can use `GetString("email_verified")`. If it's a boolean, use `GetBoolean()` before calling `ToString()`.

File: src/AspNet.Security.OAuth.DigitalOcean/DigitalOceanAuthenticationHandler.cs [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+        // PKCE https://tools.ietf.org/html/rfc7636#section-4.5, see BuildChallengeUrl
+        if (context.Properties.Items.TryGetValue(OAuthConstants.CodeVerifierKey, out var codeVerifier))
+        {
+            tokenRequestParameters.Add(OAuthConstants.CodeVerifierKey, codeVerifier!);
+            context.Properties.Items.Remove(OAuthConstants.CodeVerifierKey);
+        }
+
+        var address = QueryHelpers.AddQueryString(Options.TokenEndpoint, tokenRequestParameters);
+
+        using var request = new HttpRequestMessage(HttpMethod.Post, address);
```
Comment (martincostello) [inline src/AspNet.Security.OAuth.DigitalOcean/DigitalOceanAuthenticationHandler.cs:50]: Just checking - is it correct that all the request parameters are all in the query string and the body and headers are empty?

File: NuGet.config (hunk 1 above)
Comment (ElanHasson) [reply NuGet.config]: Oops! Sorry, was trying to get it to work locally, then i saw #633

File: src/AspNet.Security.OAuth.DigitalOcean/DigitalOceanAuthenticationHandler.cs (hunk 6 above)
Comment (ElanHasson) [reply src/AspNet.Security.OAuth.DigitalOcean/DigitalOceanAuthenticationHandler.cs]: yes, stale, great call out

File: src/AspNet.Security.OAuth.DigitalOcean/DigitalOceanAuthenticationHandler.cs (hunk 8 above)
Comment (ElanHasson) [reply src/AspNet.Security.OAuth.DigitalOcean/DigitalOceanAuthenticationHandler.cs:50]: Yeah, it's strange, we don't have to add anything per:  https://www.digitalocean.com/community/tutorials/how-to-use-oauth-authentication-with-digitalocean-as-a-user-or-developer for details

I will talk to the team internally there.

Disclaimer: I work @DigitalOcean.

File: src/AspNet.Security.OAuth.DigitalOcean/DigitalOceanAuthenticationHandler.cs (hunk 7 above)
Comment (ElanHasson) [reply src/AspNet.Security.OAuth.DigitalOcean/DigitalOceanAuthenticationHandler.cs]: `email_verified` is if the user's email address has been verified or not. It's really the only other "user" info.

The rest of the account info is related to account limits/billing/etc

Comment (martincostello) [reply src/AspNet.Security.OAuth.DigitalOcean/DigitalOceanAuthenticationHandler.cs]: My question was more about the type. `ToString()` just forces whatever it is into a string.

For example if it were an object, then it'd return `{"orly?":"yas"}`. If it's a boolean `true`, using `GetBoolean()` on it would throw if value wasn't a boolean.

Where as if it was a string of `"true"`, then `GetString("email_verified")` can be used as it's shorthand for `GetProperty("email_verified").GetString()`.

Comment (ElanHasson) [reply src/AspNet.Security.OAuth.DigitalOcean/DigitalOceanAuthenticationHandler.cs]: Makes sense.  It's a `bool`, should I still `ToLowerInvariant()`?

Comment (ElanHasson) [reply src/AspNet.Security.OAuth.DigitalOcean/DigitalOceanAuthenticationHandler.cs]: `identity.AddClaim(new Claim(DigitalOceanAuthenticationConstants.Claims.EmailVerified, account.GetProperty("email_verified").GetBoolean().ToString().ToLowerInvariant()));`

File: src/AspNet.Security.OAuth.DigitalOcean/DigitalOceanAuthenticationHandler.cs (hunk 8 above)
Comment (ElanHasson) [reply src/AspNet.Security.OAuth.DigitalOcean/DigitalOceanAuthenticationHandler.cs:50]: Spoke to a dev on that team, he pointed me to https://docs.digitalocean.com/reference/api/oauth-api/

File: src/AspNet.Security.OAuth.DigitalOcean/DigitalOceanAuthenticationHandler.cs (hunk 7 above)
Comment (martincostello) [reply src/AspNet.Security.OAuth.DigitalOcean/DigitalOceanAuthenticationHandler.cs]: I think lowering it should be fine 👍

File: src/AspNet.Security.OAuth.DigitalOcean/DigitalOceanAuthenticationHandler.cs (hunk 8 above)
Comment (martincostello) [reply src/AspNet.Security.OAuth.DigitalOcean/DigitalOceanAuthenticationHandler.cs:50]: Looking at the doc [here](https://docs.digitalocean.com/reference/api/api-reference/#operation/get_user_information), seems like `/v2/account` is effectively the "user" API here. Maybe just slightly confusing because it says "account" in the path.

Comment (martincostello) [reply src/AspNet.Security.OAuth.DigitalOcean/DigitalOceanAuthenticationHandler.cs:50]: Sorry - just realised I was answering a completely different question 😄

Comment (ElanHasson) [reply src/AspNet.Security.OAuth.DigitalOcean/DigitalOceanAuthenticationHandler.cs:50]: :D I'm told better oAuth support is coming in the future.

**Comment by ElanHasson** [comment]
@martincostello  this is ready for you.

File: src/AspNet.Security.OAuth.DigitalOcean/DigitalOceanAuthenticationConstants.cs [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+    /// <para>https://docs.digitalocean.com/reference/api/oauth-api/</para>
+    /// </summary>
+    public static class ResponseTypes
+    {
+        /// <summary>
+        /// Web Application Flow.
+        /// <para>Default scope</para>
+        /// </summary>
+        public const string Code = "code";
+    }
```
Comment (martincostello) [inline src/AspNet.Security.OAuth.DigitalOcean/DigitalOceanAuthenticationConstants.cs]: What's the motivation for adding these constants?

They don't seem to be used in the provider, so I'd be a bit reticent to add them in if they're not needed, particularly if there's plans in the works to change how DigitalOcean's OAuth implementation works.

Comment (ElanHasson) [reply src/AspNet.Security.OAuth.DigitalOcean/DigitalOceanAuthenticationConstants.cs]: I was just following the pattern and to also give some guidance to consumers as the docs come up a bit short.

You're right-- I forgot to consume them ☕

Comment (ElanHasson) [reply src/AspNet.Security.OAuth.DigitalOcean/DigitalOceanAuthenticationConstants.cs]: I appreciate your thoroughness in this PR. I based my implementation on the QQ provider, which doesn't follow the same standards.

I'm removing the `ResponseTypes`.

**Comment by martincostello** [comment]
I have to ask the question with new providers (as we've been burned in the past 😅), just to confirm, you've validated all these changes against the real DigitalOcean endpoints and have been able to successfully log into an ASP.NET Core app with it?

Other than that, are the any other additions you're planning to push up, or are you done?

**Comment by ElanHasson** [comment]
Yes, live demo: https://account.nullreference.io/ 

I am done for today :)

**Review by martincostello — APPROVED** [review]

**Comment by martincostello** [comment]
Thanks for contributing this change @ElanHasson 🎈

**Comment by ElanHasson** [comment]
Thanks for making it a pleasure!

When can I expect the NuGet to be published?

**Comment by martincostello** [comment]
I'll look at publishing it and our new provider for Xero tomorrow.

**Comment by martincostello** [comment]
Until then, you can use it from our MyGet feed: https://www.myget.org/feed/aspnet-contrib/package/nuget/AspNet.Security.OAuth.DigitalOcean

**Comment by ElanHasson** [comment]
Thanks!
